File: backend/open_webui/utils/openai_provider_urls.py
"""Provider-aware URL and model-id helpers for OpenAI-compatible upstreams."""

from urllib.parse import urlparse


def resolve_openai_chat_completions_upstream_url(base_url: str) -> str:
    """
    Return the POST URL for chat (streaming or not).

    Free.ai documents ``POST /v1/chat/`` (with trailing slash) instead of
    OpenAI's ``/v1/chat/completions``. Same request/response shape as chat
    completions; only the path differs.
    """
    base = base_url.rstrip('/')
    host = (urlparse(base).hostname or '').lower()
    if host == 'api.free.ai':
        return f'{base}/chat/'
    return f'{base}/chat/completions'


def resolve_openai_provider_model_id(base_url: str, model_id: str, prefix_id: str | None = None) -> str:
    """
    Normalize model IDs before sending upstream.

    - strips configured ``prefix_id.`` decoration (Open WebUI display prefix)
    - applies vendor-specific normalization where needed
    """
    model = (model_id or '').strip()
    if not model:
        return model

    if prefix_id:
        model = model.removeprefix(f'{prefix_id}.')

    host = (urlparse((base_url or '').rstrip('/')).hostname or '').lower()
    if host == 'integrate.api.nvidia.com':
        # NVIDIA model ids are vendor/model (slash), not dotted display ids.
        if model.startswith('nvidia.'):
            model = model.removeprefix('nvidia.')
        if '/' not in model and '.' in model:
            left, right = model.split('.', 1)
            model = f'{left}/{right}'

    return model

File: src/lib/utils/openaiProviderUrls.ts
/** Provider-aware chat/model normalization for OpenAI-compatible upstreams. */

export function resolveOpenAIChatCompletionsUrl(baseUrl: string): string {
	const base = baseUrl.replace(/\/+$/, '');
	try {
		const host = new URL(base).hostname.toLowerCase();
		if (host === 'api.free.ai') {
			return `${base}/chat/`;
		}
	} catch {
		// invalid baseUrl — fall through
	}
	return `${base}/chat/completions`;
}

export function resolveOpenAIProviderModelId(
	baseUrl: string,
	modelId: string,
	prefixId?: string | null
): string {
	let model = (modelId ?? '').trim();
	if (!model) return model;

	if (prefixId) {
		model = model.replace(new RegExp(`^${prefixId.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\.`), '');
	}

	try {
		const host = new URL(baseUrl.replace(/\/+$/, '')).hostname.toLowerCase();
		if (host === 'integrate.api.nvidia.com') {
			if (model.startsWith('nvidia.')) {
				model = model.slice('nvidia.'.length);
			}
			if (!model.includes('/') && model.includes('.')) {
				const [left, right] = model.split('.', 2);
				model = `${left}/${right}`;
			}
		}
	} catch {
		// Ignore malformed URL and return best-effort model id.
	}

	return model;
}
